File: common/src/main/java/juuxel/adorn/block/BenchBlock.java
package juuxel.adorn.block;

import juuxel.adorn.lib.AdornStats;
import juuxel.adorn.util.AdornUtil;
import juuxel.adorn.util.ShapeRotation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.resources.Identifier;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.util.RandomSource;
import net.minecraft.world.phys.shapes.VoxelShape;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;

import java.util.ArrayList;
import java.util.List;

public final class BenchBlock extends SeatBlock implements SimpleWaterloggedBlock, BlockWithDescription {
    public static final EnumProperty<Direction.Axis> AXIS = BlockStateProperties.HORIZONTAL_AXIS;
    public static final BooleanProperty CONNECTED_N = BooleanProperty.create("connected_n");
    public static final BooleanProperty CONNECTED_P = BooleanProperty.create("connected_p");
    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;

    private static final String DESCRIPTION_KEY = "block.adorn.bench.description";
    private static final VoxelShape X_TOP_SHAPE = box(0.0, 8.0, 1.0, 16.0, 10.0, 15.0);
    private static final VoxelShape Z_TOP_SHAPE = box(1.0, 8.0, 0.0, 15.0, 10.0, 16.0);
    private static final VoxelShape[] SHAPES = new VoxelShape[8];

    static {
        var legShapes = ShapeRotation.buildShapeRotationsFromNorth(2, 0, 2, 14, 8, 4);
        var booleans = new boolean[] { true, false };

        for (var axis : new Direction.Axis[] { Direction.Axis.X, Direction.Axis.Z }) {
            var topShape = axis == Direction.Axis.X ? X_TOP_SHAPE : Z_TOP_SHAPE;
            var negativeLeg = legShapes.get(Direction.fromAxisAndDirection(axis, Direction.AxisDirection.NEGATIVE));
            var positiveLeg = legShapes.get(Direction.fromAxisAndDirection(axis, Direction.AxisDirection.POSITIVE));

            for (var connectedN : booleans) {
                for (var connectedP : booleans) {
                    List<VoxelShape> parts = new ArrayList<>();

                    if (!connectedN) {
                        parts.add(negativeLeg);
                    }

                    if (!connectedP) {
                        parts.add(positiveLeg);
                    }

                    int key = getShapeKey(axis, connectedN, connectedP);
                    SHAPES[key] = Shapes.or(topShape, parts.toArray(VoxelShape[]::new));
                }
            }
        }
    }

    public BenchBlock(Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState()
            .setValue(AXIS, Direction.Axis.Z)
            .setValue(CONNECTED_N, false)
            .setValue(CONNECTED_P, false)
            .setValue(WATERLOGGED, false));
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        var state = defaultBlockState()
            .setValue(AXIS, AdornUtil.turnHorizontally(ctx.getHorizontalDirection().getAxis()))
            .setValue(WATERLOGGED, ctx.getLevel().getFluidState(ctx.getClickedPos()).getType() == Fluids.WATER);
        return updateConnections(ctx.getLevel(), ctx.getClickedPos(), state);
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        if (state.getValue(WATERLOGGED)) {
            tickView.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(world));
        }

        return updateConnections(world, pos, state);
    }

    private BlockState updateConnections(BlockGetter world, BlockPos pos, BlockState state) {
        for (var axisDirection : Direction.AxisDirection.values()) {
            var property = axisDirection == Direction.AxisDirection.NEGATIVE ? CONNECTED_N : CONNECTED_P;
            var neighbor = world.getBlockState(pos.relative(state.getValue(AXIS), axisDirection.getStep()));
            var connected = neighbor.getBlock() instanceof BenchBlock && neighbor.getValue(AXIS) == state.getValue(AXIS);
            state = state.setValue(property, connected);
        }
        return state;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return SHAPES[getShapeKey(state.getValue(AXIS), state.getValue(CONNECTED_N), state.getValue(CONNECTED_P))];
    }

    private static int getShapeKey(Direction.Axis axis, boolean connectedN, boolean connectedP) {
        return (axis == Direction.Axis.X ? 1 : 0) << 2 | (connectedN ? 1 : 0) << 1 | (connectedP ? 1 : 0);
    }

    @Override
    public FluidState getFluidState(BlockState state) {
        return state.getValue(WATERLOGGED) ? Fluids.WATER.getSource(false) : super.getFluidState(state);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(AXIS, CONNECTED_N, CONNECTED_P, WATERLOGGED);
    }

    @Override
    public BlockState rotate(BlockState state, Rotation rotation) {
        return switch (rotation) {
            case COUNTERCLOCKWISE_90:
            case CLOCKWISE_90:
                yield state.setValue(AXIS, AdornUtil.turnHorizontally(state.getValue(AXIS)));
            default:
                yield state;
        };
    }

    @Override
    public Identifier getSittingStat() {
        return AdornStats.SIT_ON_BENCH;
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    public Direction getPreferredDismountDirection(BlockState state, Entity passenger) {
        var axis = AdornUtil.turnHorizontally(state.getValue(AXIS));
        var passengerFacing = passenger.getDirection();
        return passengerFacing.getAxis() == axis ? passengerFacing : Direction.get(Direction.AxisDirection.POSITIVE, axis);
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/BenchBlock.kt
package juuxel.adorn.block

import it.unimi.dsi.fastutil.bytes.Byte2ObjectArrayMap
import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap
import juuxel.adorn.api.block.BlockVariant
import juuxel.adorn.lib.AdornStats
import juuxel.adorn.util.buildShapeRotationsFromNorth
import juuxel.adorn.util.getDirection
import juuxel.adorn.util.turnHorizontally
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.Waterloggable
import net.minecraft.fluid.FluidState
import net.minecraft.fluid.Fluids
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.EnumProperty
import net.minecraft.state.property.Properties
import net.minecraft.util.BlockRotation
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.WorldAccess

class BenchBlock(variant: BlockVariant) : SeatBlock(variant.createSettings()), Waterloggable, BlockWithDescription {
    override val sittingStat = AdornStats.SIT_ON_BENCH
    override val descriptionKey = "block.adorn.bench.description"

    init {
        defaultState = defaultState
            .with(AXIS, Direction.Axis.Z)
            .with(CONNECTED_N, false)
            .with(CONNECTED_P, false)
            .with(WATERLOGGED, false)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState =
        defaultState
            .with(AXIS, ctx.playerFacing.axis.turnHorizontally())
            .with(WATERLOGGED, ctx.world.getFluidState(ctx.blockPos).fluid == Fluids.WATER)
            .let { state -> updateConnections(ctx.world, ctx.blockPos, state) }

    override fun getStateForNeighborUpdate(
        state: BlockState, direction: Direction, neighborState: BlockState,
        world: WorldAccess, pos: BlockPos, neighborPos: BlockPos
    ): BlockState {
        if (state[WATERLOGGED]) {
            world.createAndScheduleFluidTick(pos, Fluids.WATER, Fluids.WATER.getTickRate(world))
        }

        return updateConnections(world, pos, state)
    }

    private fun updateConnections(world: BlockView, pos: BlockPos, state: BlockState): BlockState =
        Direction.AxisDirection.values().fold(state) { acc, axisDirection ->
            val property = if (axisDirection == Direction.AxisDirection.NEGATIVE) CONNECTED_N else CONNECTED_P
            val neighbor = world.getBlockState(pos.offset(state[AXIS], axisDirection.offset()))
            val connected = neighbor.block is BenchBlock && neighbor[AXIS] == state[AXIS]

            acc.with(property, connected)
        }

    override fun getOutlineShape(state: BlockState, world: BlockView, pos: BlockPos, ctx: ShapeContext): VoxelShape =
        SHAPES[Bits.buildBenchState(state[AXIS], state[CONNECTED_N], state[CONNECTED_P])]

    @Suppress("DEPRECATION")
    override fun getFluidState(state: BlockState): FluidState =
        if (state[Properties.WATERLOGGED]) {
            Fluids.WATER.getStill(false)
        } else {
            super.getFluidState(state)
        }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(AXIS, CONNECTED_N, CONNECTED_P, WATERLOGGED)
    }

    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState =
        if (rotation == BlockRotation.COUNTERCLOCKWISE_90 || rotation == BlockRotation.COUNTERCLOCKWISE_90) {
            state.with(AXIS, state[AXIS].turnHorizontally())
        } else {
            state
        }

    companion object {
        val AXIS: EnumProperty<Direction.Axis> = Properties.HORIZONTAL_AXIS
        val CONNECTED_N: BooleanProperty = BooleanProperty.of("connected_n")
        val CONNECTED_P: BooleanProperty = BooleanProperty.of("connected_p")
        val WATERLOGGED: BooleanProperty = Properties.WATERLOGGED

        private val X_TOP_SHAPE: VoxelShape = createCuboidShape(0.0, 8.0, 1.0, 16.0, 10.0, 15.0)
        private val Z_TOP_SHAPE: VoxelShape = createCuboidShape(1.0, 8.0, 0.0, 15.0, 10.0, 16.0)
        private val SHAPES: Byte2ObjectMap<VoxelShape> = Byte2ObjectArrayMap()

        init {
            val legShapes = buildShapeRotationsFromNorth(2, 0, 2, 14, 8, 4)
            val booleans = booleanArrayOf(true, false)

            for (axis in arrayOf(Direction.Axis.X, Direction.Axis.Z)) {
                val topShape = if (axis == Direction.Axis.X) X_TOP_SHAPE else Z_TOP_SHAPE
                val negativeLeg = legShapes[axis.getDirection(Direction.AxisDirection.NEGATIVE)]!!
                val positiveLeg = legShapes[axis.getDirection(Direction.AxisDirection.POSITIVE)]!!

                for (connectedN in booleans) {
                    for (connectedP in booleans) {
                        val parts = ArrayList<VoxelShape>()

                        if (!connectedN) {
                            parts += negativeLeg
                        }

                        if (!connectedP) {
                            parts += positiveLeg
                        }

                        val key = Bits.buildBenchState(axis, connectedN, connectedP)
                        SHAPES[key] = VoxelShapes.union(topShape, *parts.toTypedArray())
                    }
                }
            }
        }
    }
}
